src/compiler/utils/string_with_sourcemap.ts

```diff
@@ -13,21 +13,22 @@ function last_line_length(s: string) {
 
 // mutate map in-place
 export function sourcemap_add_offset(
-	map: DecodedSourceMap, offset: SourceLocation
+	map: DecodedSourceMap, offset: SourceLocation, source_index: number
 ) {
-	if (map.mappings.length == 0) return map;
-	// shift columns in first line
-	const segment_list = map.mappings[0];
-	for (let segment = 0; segment < segment_list.length; segment++) {
-		const seg = segment_list[segment];
-		if (seg[3]) seg[3] += offset.column;
-	}
-	// shift lines
+	if (map.mappings.length == 0) return;
 	for (let line = 0; line < map.mappings.length; line++) {
 		const segment_list = map.mappings[line];
 		for (let segment = 0; segment < segment_list.length; segment++) {
 			const seg = segment_list[segment];
-			if (seg[2]) seg[2] += offset.line;
+			// shift only segments that belong to component source file
+			if (seg[1] === source_index) { // also ensures that seg.length >= 4
+				// shift column if it points at the first line
+				if (seg[2] === 0) {
+					seg[3] += offset.column;
+				}
+				// shift line
+				seg[2] += offset.line;
+			}
```

Comment:
would be nice to have a test for "only shift lines of one source file"
probably you have good reasons for this change, but its not yet transparent

if your solution is right, then we get another problem:
what about basename collisions?
`a/index.svelte + b/index.svelte --> c/index.svelte`

the sourcemap.names array is not sorted, so its not predictable
so there is no (simple) way to choose the right `index.svelte` file
lazy solution: print a warning.
"dear user, sorry, but our compiler knows files only by their basename .."

a more radical solution: instead of file names, use file paths.
these are always relative to the project root
all our plugins would have to preserve these paths

Reply:
> test for "only shift lines of one source file"

`sourcemap-offsets` should cover this. Or am i misunderstanding you?

> what about basename collisions?

I think that I have a different vision of the build pipeline:

1. Preprocess
- We assume that preprocess src === dst and map file is placed in same dir. We use `basename` for component in sources since it should be relative to the sourcemap file.
- We use `basename` for chunks that we don't pass to preprocessors (`StringWithSourcemap.from_source`)
- We pass full `filename` to preprocessors and expect sources to contain `basename` when referring component being preprocessed and `relative to filename paths` for external inclusions. For example, expect `["index.svelte", "../b/index.svelte"]` when transforming `a/index.svelte` (no collision here).

2. Compile
- We pass `filename`, `outputFilename` and `cssOutputFilename`. Based on these, svelte compiler creates source map with sources as relative paths.
- Then we combine js and css maps from compiler with preprocessed map so that segments point at locations in original source code and sources in resulted maps are relative from output map file path to original source files.

3. Bundle (optional)
- Should work as usual since we input compiled files with maps pointing directly at original sources.

Is there anything missing, you disagree with or that I should describe more clearly? I'd prefer not to land something cryptic or/and broken :)

P.S. Just out of interest, is it synthetic or do you know setups when preprocess step combines multiple components into one?

Reply:
> `sourcemap-offsets` should cover this.

yes! sorry. ("dear mila, please just read the source")

> We use `basename` for component in sources since it should be relative to the sourcemap file.

yes .. i wrongly assumed that the basename-function is applied to the preprocessor-result
while its only applied to the svelte.preprocess input filename

> We pass full `filename` to preprocessors and expect sources to contain `basename` when referring component being preprocessed

this is my problem with this solution ..
why do we **always** convert file paths to basenames?
(always, without giving an option to the user)

short answer: for consistency with the existing svelte.compile

good news: i am happy with your patch : )

long answer:

<details>

rollup@2.34.2 + svelte@3.31.0
svelte.preprocess and svelte.compile receive relative paths
`options = { filename: 'src/App.svelte' }`

webpack@4.30.0 + svelte@3.31.0
svelte.preprocess and svelte.compile receive absolute paths
`options = { filename: '/home/user/project/src/App.svelte' }`

> why do we **always** convert file paths to basenames?

the actual reason here is MagicString, who turns file paths into basenames:

```js
// svelte/src/compiler/compile/render_dom/index.ts
const css = component.stylesheet.render(
  options.filename, // file path
  !options.customElement
);

// svelte/src/compiler/compile/css/Stylesheet.ts
// Stylesheet.render
return {
  code: code.toString(),
  map: code.generateMap({
    includeContent: true,
    source: this.filename, // file path
    file
  })
};
// --> result.map.sources: basename

// magic-string/src/MagicString.js
// MagicString.generateDecodedMap
return {

  // path to basename
  file: options.file ? options.file.split(/[/\\]/).pop() : null,

  sources: [options.source ? getRelativePath(options.file || '', options.source) : null],
  sourcesContent: options.includeContent ? [this.original] : [null],
  names,
  mappings: mappings.raw
};
```

so the path-to-basename conversion is done for consistency with other sourcemaps, generated by svelte.compile

</details>

Reply:
> why do we always convert file paths to basenames?

`basename` is just a result of computing `relative path` from file to itself since I assumed that we "put" results of preprocess into the same directory where source code lives.

If someone needs more control we could add additional option `outputFilename` to preprocess so that specific relative paths are generated instead. For example, take `./src/index.svelte`, preprocess it and put into `./preprocessed/index.svelte`, and then pass those files to svelte.compile independently. In this case, instead `index.svelte` (basename) we would have our component referred by index.svelte.map as `../src/index.svelte` (relative to map file location - from `{cwd}/preprocessed/index.svelte.map` to `{cwd}/src/index.svelte`).

But I'm not sure that such feature is really needed and worth complication both in usage and implementation since usually preprocess is done in-memory and can be treated as atomic operation with compile. I might be wrong here though. But then it can be implemented separately.

So I would say it's more about simplicity, not consistency.

Reply:
can we add a test for the basename conversion? something like ....


<details>

```js
import { magic_string_bundle } from '../../helpers';

const external_filename = 'external_code.css';
export const external_code = `

span {
	--external_code-var: 1px;
}

`;

const component_filepath = 'src/input.svelte';
const component_basename = 'input.svelte';

export default {
	js_map_sources: [component_basename],
	css_map_sources: [component_basename, external_filename],
	preprocess: [
		{
			style: ({ content, filename }) => {
				return magic_string_bundle([
					{ code: external_code, filename: external_filename },
					{ code: content, filename }
				]);
			}
		}
	],
	options: {
		filename: component_filepath
	}
};
```

</details>

> it can be implemented separately.

yes .. low priority stuff

what i meant was:
optionally, svelte should "simply" preserve all file paths (relative or absolute)
then its the bundler's job (rollup, webpack) to feed pretty file paths to svelte

for example

`src/App.svelte` should be preserved

when `src/App.svelte` includes `./style/included.css` 
then its resolved to `src/style/included.css`

makes easier the debugging of larger projects with many files + folders

Reply:
thx : )

Reply:
> makes easier the debugging of larger projects with many files + folders

Sorry, will response ASAP.

Reply:
well thats easy ..
https://github.com/milahu/svelte-template--reproduce-basename-vs-filepath

actual result:
all paths are reduced to basename
(see my other comment, this is done by MagicString.generateDecodedMap)

<img src="https://user-images.githubusercontent.com/12958815/101993781-961e8480-3cbd-11eb-849e-ed8120a4acff.png" width="400" />

expeted result:
optionally show the full file path, for example `src/menu/App.svelte`

Reply:
I made the filename relative to cwd in `rollup-plugin-svelte` in https://github.com/sveltejs/rollup-plugin-svelte/pull/131 because that's what I had seen in the couple example source maps I looked at. Assuming that's the correct behavior (I haven't fully followed the full discussion here), we could make that change to webpack's `svelte-loader` as well.

Reply:
I cloned it, launched and console logged the same. I also did some tweak after, will describe in my next comment. But if I hover App.svelte:2, I see full url to correct source file. Clicking also navigates to where I would expect.